### PR TITLE
e2e Holo Hosting Tests - B 02101

### PR DESCRIPTION
**Pull request by JettTech**
### Updates:
 - adds puppeteer setup for hosted agent browser simulation
 - adds tryorama setup for holo hosted players and holochain players
 - adds sign up/sign out test (auth test)
 - adds update profile test (zome call test to agent's chain)
 - adds create request test (zome call test with counterparty)

### How to run e2e Holo Hosting Tests
This runs with a version of Tryorama that includes the consistency updates and hosted player additoin (branch `2020-08-31/hosted-consistency`). **Before running, you will need to pull down a local copy and point the tryorama dependency path to your local copy.**  (In `package.json`, simply replace the version number with `file:/path/to/local/tryorama/dir`)

 > Pro Tips: 
>- If the test fails to successfully open a tab, make sure the UI server is running (198.211.102.13:3000).
>- If the test fails prior to completing sign-in, cancel the running test and rerun.
>- If the test hangs after signing in, make sure the chaperone server is running (by default this is connected to the test chaperone server noted in `holochainClient.js`).

 - Verify that a non-corrupted hpos is up and registered as a host for the happ (`QmNedTibHaD3K7ojqa7ZZfkMBbUWg39taK6oLBEPAswTKu`) in the resolver-dev KV store (`250ood7t7wevast9gmkdn91eg9r026o82itj2z50q6ljhctc43` works well).
 - run `npm i` to make sure you've installed all the lasted deps (notably the new @holo-host/web-sdk)
 - enter `nix-shell` (NB: This is needed for the tryorama instances.)
 - run `npm run test:integration-holo`

 **Testing Details:**
   - Puppeteer is used to emulate the Holo Hosted Player actions.
   - The Holochain Player is spawned and managed by Tryorama.
   - A Hosted Player is also created in Tryorama to monitor consistency. 
   - Consistency is managed by connecting to the Host of the Holo Hosted User and contrasting the Hosted Agent's Instance state dump with that of the Holochain Instance to determine if both the Hosted Player and the Holochain Player agree on the current state (ie: `simpleConsistency`).

 #### Sign Up / Sign out Test (Web SDK Test)
 - This test emulates a Holo Hosted User signing up and signing out successfully.
  - This test is considered successful once the Hosted Player's signs up, arrives at the HoloFuel Inbox page and is able to sign out, arriving back at the sign in modal.

 #### Update ProfileTest (Zome Call Test)
 - This test emulates a Holo Hosted User signing up, navigating to the profile page, and updating his/her agent nickname.
  - This test is considered successful once the Hosted Player's new nickname is discoverable by the holochain player.

 #### Send Request Test (Zome Call Test)
 - This test emulates a Holo Hosted User signing up, navigating to the number pad/create request page, and initiating a request for HoloFuel from a Holochain instance.
  - This test is considered successful once the Hosted Player's request is found in the counterparty's (the holochain player's) list of pending transactions.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,9 +35,11 @@
     "preseed:holofuel": "node scripts/prepareHoloFuelData/index.js",
     "preseed:hpadmin": "node scripts/prepareHpAdminData/index.js",
     "test:preseed-data": "npm run test:wait-for-conductor && NODE_ENV=test babel-node scripts/initialize-conductor-data",
-    "test:integration": "nix-shell --command 'REACT_APP_INTEGRATION_TEST=true node scripts/test.js --config=./jest.integration.config.js'",
+    "test:integration-holo": "REACT_APP_INTEGRATION_TEST='true' node scripts/test.js --config=./jest.holoIntegration.config.js --runInBand --testTimeout=95000000",
+    "test:integration-holochain": "nix-shell --command 'REACT_APP_INTEGRATION_TEST='true' node scripts/test.js --config=./jest.hcIntegration.config.js'",
     "test:integration-with-setup": "npm run test:preseed-data && npm run test:integration",
-    "predeploy": "npm run build:holofuel",
+    "serve-test": "PORT=3100 npm run start:holofuel-holo-live",
```
Comment (robbiecarlton) [inline package.json]: can we remove `test:integration-with-setup ` command now? Or is that still relevant?

File: package.json [hunk 2]
```diff
@@ -44,8 +49,9 @@
     "@babel/node": "^7.6.3",
     "@holo-host/chaperone": "^0.2.0",
     "@holo-host/hp-admin-keypair": "0.1.2",
-    "@holo-host/web-sdk": "0.1.2",
+    "@holo-host/web-sdk": "^0.1.2",
     "@holochain/hc-web-client": "^0.5.3",
+    "@holochain/tryorama": "file:/home/lisa/Documents/gitrepos/holochain/holochain-old/tryorama",
```
Comment (robbiecarlton) [inline package.json]: Can this be replaced with current Tryorama? Can't merge with this.

File: nixpkgs.nix [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import (fetchTarball {
-  url = "https://github.com/Holo-Host/holo-nixpkgs/archive/8b122836e8074c289abe49b80755e614125fb820.tar.gz";
-  sha256 = "1fnsnp5g7ylr8b1qpn0fgrk2pw477aah0s0s644svjd24qri62ks";
+  # commit: 6506590fc254932b4a3b9ba5c9f965c042ab2941
```
Comment (robbiecarlton) [inline nixpkgs.nix]: remove comment

File: .env.example [hunk 4]
```diff
@@ -1,6 +1,7 @@
 REACT_APP_DNA_INTERFACE_URL=ws://localhost:33000
 REACT_APP_HPOS_URL=ws://localhost:3500
 REACT_APP_AGENT_ID=HcSCjCvkR5uywn6z5t9SBFtTs7p83Ex3urUtpY8jsUa7pn3zdF6XJU4U6tuzpui
-REACT_APP_DEFAULT_STORAGE='/[Path]/[to]/[this]/[repo]/.holochain/holo/storage'
-REACT_APP_STORAGE_SNAPSHOT='/[Path]/[to]/[this]/[repo]/.holochain/holo/storageSnapshot'
 REACT_APP_VERSION=$npm_package_version
+
+REACT_APP_TEST_DNA_INSTANCE_ID=holofuel-dna-instance
+REACT_APP_DNA_PATH=/[Path]/[to]/[holofuel]/[repo]/.dna.json
```
Comment (robbiecarlton) [inline .env.example]: Doesn't look like you're using this env variable anymore

File: package.json [hunk 5]
```diff
@@ -17,13 +17,16 @@
     "start:conductor-2": "HC_IGNORE_SIM2H_URL_PROPERTY=true holochain -c conductor-config-2.toml",
     "start:sim2h-server": "sim2h_server -p 9000",
     "start:dev-chaperone": "npx chaperone-server --config chaperone.json",
+    "sim2h-server:start": "sim2h_server -p 9000 &> sim2h.log &",
+    "sim2h-server:clear": "kill -9 $(lsof -t -i:9000)",
+    "refresh": "npm run hc:stop; npm run sim2h-server:clear; npm run hc:clean-storage",
```
Comment (robbiecarlton) [inline package.json]: I find `refresh` a little misleading. Maybe `stop-all`?

File: scripts/wait-for-conductor.js [hunk 6]
```diff
@@ -38,7 +38,7 @@ async function waitForConductor (interval = 10000) {
         console.log('')
         return wait(interval)
       } else {
-        console.log('>>>>>> CALL TO CONDUCTOR DID NOT PASS :: error code : ', error.code)
+        // console.log('>>>>>> CALL TO CONDUCTOR DID NOT PASS :: error code : ', error.code)
```
Comment (robbiecarlton) [inline scripts/wait-for-conductor.js]: is this still supposed to be commented out?

File: src/holofuel/components/layout/PrimaryLayout/PrimaryLayout.module.css [hunk 7]
```diff
@@ -45,4 +45,5 @@
   font-size: 20px;
   margin: 0 auto;
   margin-top: 75px;
-}
\ No newline at end of file
+}
+/* connection text #389ea6 */
```
Comment (robbiecarlton) [inline src/holofuel/components/layout/PrimaryLayout/PrimaryLayout.module.css]: remove comment

File: holoIntegration-setup/setup.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+// setup.js
+const puppeteer = require('puppeteer');
+const mkdirp = require('mkdirp');
+const path = require('path');
+const fs = require('fs');
+const os = require('os');
+
+const HAPP_URL = 'http://testfuel.holo.host/'
```
Comment (robbiecarlton) [inline holoIntegration-setup/setup.js]: This variable isn't being used

File: holoIntegration-setup/teardown.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+// teardown.js
+const os = require('os');
+const rimraf = require('rimraf');
+const path = require('path');
+
+const DIR = path.join(os.tmpdir(), 'jest_puppeteer_global_setup');
```
Comment (robbiecarlton) [inline holoIntegration-setup/teardown.js]: Would be more robust to import this from the other file.

**Review by robbiecarlton — CHANGES_REQUESTED** [review]
A couple of code cleanup changes here.

But also, 

`test:integration-holo` and `test:integration-holochain` are failing for me locally, and tests and linter are failing on CI.

File: package.json (hunk 2 above)
Comment (JettTech) [reply package.json]: This currently needs to be run with a specific build of tryorama (off of branch`2020-08-31/hosted-consistency`).  Until that PR is merged, you will need to pull down a local copy and point to your local path.

**Comment by JettTech** [comment]
NOTE: Locally, with a copy of tryorama built off of branch`2020-08-31/hosted-consistency`, both linting and tests pass. We wll need to await [Tryorama PR #36](https://github.com/holochain/tryorama/pull/36) to be merged in order for CI to properly run tests.

**Comment by JettTech** [comment]
@robbiecarlton - This PR is *only* for e2e holo hosting tests, not holochain.  

Also, please read the updated description for how to run to the tests, and let me know if you still run into issues running the holo integration tests

File: package.json (hunk 1 above)
Comment (JettTech) [reply package.json]: We can.  This PR does not work on the holochain integration tests, so the work on those remains untouched in this version. That said, this script is deprecated, so I am on board with removing it anyway.

File: src/holofuel/__integration_tests__/holo/CreateRequest.holoIntegration.test.js [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+      const submitButton = await page.$("button[data-testid='submit-button']")
+      submitButton.click()
+
+      // // wait for DHT consistency
+      await awaitSimpleConsistency(scenario, DNA_INSTANCE, [counterpartyAgentInstance], [hostedAgentInstance])
+
+      const checkListPending = async () => counterpartyAgentInstance.call('holofuel', 'transactions', 'list_pending', {})
+      const listPending = await waitZomeResult(checkListPending, 90000, 10000)
+
+      expect(listPending.requests.length).toEqual(1)
```
Comment (zippy) [inline src/holofuel/__integration_tests__/holo/CreateRequest.holoIntegration.test.js]: I think we should add a zome call check before the request that calls checkListPending and ensures that the length is 0.

**Review by zippy — APPROVED** [review]

**Review by zippy — APPROVED** [review]
This is fantastic work.  Definitely good-enough for us to build from for when we get the rsm-context working.

One tiny request is to have a "before check" on the CreateRequest test that shows proves the transition from 0 to 1 items in list pending.

Also I want to note that we are going to have to do some work with these tests that use HostedPlayer in making sure that the initial state is clean.  I would eventually like to be able to run this test from inside final-exam, but I don't think that's currently possible because there are hard references to the active holoport, i.e. they all seem to use Joel's holoport TEST_HOST[0].  I think a next step would be for the test runner to fully spawn an hpos-shell that it uses for the test, not a holoport.  But that should happen as a followup.

**Review by zo-el — APPROVED** [review]

**Review by robbiecarlton — CHANGES_REQUESTED** [review]
Tests and linter still failing on CI.

Not sure why linter is failing, tests most likely failing because of the `file:` setting for the Tryorama dependency in `package.json`.

Since we're not planning on running the E2E tests in this CI, my recommendation is that you just put in a normal version constraint for Tryorama in `package.json`, so that CI tests pass and `npm i` works when downloading the repo.

The`README.md` instructions explains how to update the Tryorama thing.

A couple of other change request:

Could you add a .gitignore file inside `src/holofuel/__integration_tests__/holo/snapshots` so that the directory gets committed?

And could you get package-lock.json back?

If you just use the one in `holo-hosted-holofuel/v0.14.X` and then it might pick up a small change from installing Tryorama.

These changes plus once I've successfully run the tests locally and we're good to go

File: src/holochainClient.js [hunk 11]
```diff
@@ -12,7 +12,7 @@ const CHAPERONE_SERVER = {
   test: 'http://198.199.73.20:8000/',
   production: 'https://chaperone.holo.host/'
 }
-const CHAPERONE_SERVER_URL = CHAPERONE_SERVER.production
+export const CHAPERONE_SERVER_URL = CHAPERONE_SERVER.test
```
Comment (robbiecarlton) [inline src/holochainClient.js]: this should branch based on an environment variable right? We don't want `CHAPERONE_SERVER_URL = CHAPERONE_SERVER.test` in production.

Might be simplest to just have a `REACT_APP_CHAPERONE_SERVER_URL`  environment variable and not have these hard coded values at all.

**Review by robbiecarlton — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 12]
```diff
@@ -18,6 +18,7 @@
 .env.development.local
 .env.test.local
 .env.production.local
+src/holofuel/__integration_tests__/holo/snapshots/*
```
Comment (robbiecarlton) [inline .gitignore]: @JettTech This doesn't mean the directory is committed. You have to add a .gitignore file to the directory you want to be committed, so `src/holofuel/__integration_tests__/holo/snapshots/` in this case.

File: src/holofuel/__integration_tests__/holo/snapshots/.gitignore [hunk 13]
```diff
@@ -0,0 +1,3 @@
+*
+
+!.gitignore
```
Comment (robbiecarlton) [inline src/holofuel/__integration_tests__/holo/snapshots/.gitignore:3]: @JettTech - adding this file means that the snapshots directory will be created when you check out this branch, but nothing in the directory (except this file itself) will be committed

File: .gitignore [hunk 14]
```diff
@@ -18,7 +18,6 @@
 .env.development.local
 .env.test.local
 .env.production.local
-src/holofuel/__integration_tests__/holo/snapshots/*
```
Comment (robbiecarlton) [inline .gitignore]: @JettTech - this line won't actually create the snapshots directory. See the `.gitignore` file I added for an example of a way to do that.

**Review by robbiecarlton — APPROVED** [review]

**Pull request merged**